osquery/events/darwin/diskarbitration.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include "osquery/core/conversions.h"
+#include "osquery/events/darwin/diskarbitration.h"
+
+namespace osquery {
+
+REGISTER(DiskArbitrationEventPublisher, "event_publisher", "diskarbitration");
+
+void DiskArbitrationEventPublisher::restart() {
+  if (run_loop_ == nullptr) {
+    return;
+  }
+  stop();
+
+  if (session_ == nullptr) {
+    session_ = DASessionCreate(kCFAllocatorDefault);
+  }
+
+  DARegisterDiskAppearedCallback(
+      session_,
+      nullptr,
+      DiskArbitrationEventPublisher::DiskAppearedCallback,
+      nullptr);
+  DARegisterDiskDisappearedCallback(
+      session_,
+      nullptr,
+      DiskArbitrationEventPublisher::DiskDisappearedCallback,
+      nullptr);
+
+  DASessionScheduleWithRunLoop(session_, run_loop_, kCFRunLoopDefaultMode);
+}
+
+void DiskArbitrationEventPublisher::DiskAppearedCallback(DADiskRef disk,
+                                                         void *context) {
+  CFDictionaryRef disk_properties = DADiskCopyDescription(disk);
+  CFTypeRef devicePathKey;
+  if (!CFDictionaryGetValueIfPresent(
+          disk_properties, kDADiskDescriptionDevicePathKey, &devicePathKey)) {
+    return;
+  }
+
+  auto device_path = stringFromCFString((CFStringRef)devicePathKey);
+
+  // Check if starts with kIOHIDXClass
+  // if it doesn't, it's not a virtual disk
+  if (!boost::starts_with(device_path, kIOHIDXClassPath_)) {
```

Comment:
What happens if you don't include this restriction, will we see physical mounts too? That might be a good thing!